### PR TITLE
Add prometheus metrics for DataNode

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: #15592

### CODE REVIEW

File: internal/metrics/datanode.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+			"num_flow_graphs",
+			nodeIDLabelName,
+		})
+
+	DataNodeConsumeMsgRowsCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "message_size_count",
+			Help:      "Messages size count consumed from msgStream in DataNode.",
```
Comment (czs007) [inline internal/metrics/datanode.go:70]: rows count consumed from msgStream in DataNode

File: internal/metrics/datanode.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+		})
+
+	DataNodeNumDeltaChannels = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_delta_channels",
+			Help:      "Number of deltaChannels per collection in DataNode.",
+		}, []string{
+			"num_delta_channels",
```
Comment (czs007) [inline internal/metrics/datanode.go:105]: add collection_id label to distinguish different collections.

File: internal/metrics/datanode.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		})
+
+	DataNodeNumConsumers = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_consumers",
+			Help:      "Number of consumers per collection in DataNode.",
+		}, []string{
+			"num_consumers",
```
Comment (czs007) [inline internal/metrics/datanode.go:116]: add colleciton_id to distinguish different collections.

File: internal/metrics/datanode.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+			nodeIDLabelName,
+		})
+
+	DataNodeNumProducers = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_producers",
+			Help:      "Number of producers per collection in DataNode.",
```
Comment (czs007) [inline internal/metrics/datanode.go:125]: add collection_id label to distinguish different collections.

File: internal/metrics/datanode.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		})
+
+	DataNodeTimeSync = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "time_sync_per_channel",
+			Help:      "Synchronized timestamps per channel in DataNode.",
+		}, []string{
+			"time_sync_per_channel",
```
Comment (czs007) [inline internal/metrics/datanode.go:138]: add pchan label to distinguish different physical channels.

File: internal/metrics/datanode.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_unflushed_segments",
+			Help:      "Number of unflushed segments in DataNode.",
+		}, []string{
+			"num_unflushed_segments",
+			nodeIDLabelName,
+		})
+
+	DataNodeFlushDuration = prometheus.NewHistogramVec(
```
Comment (czs007) [inline internal/metrics/datanode.go:164]: his refers to the time interval between the datanode receiving the flush request and the completion of the flush operation.

File: internal/metrics/datanode.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "save_duration",
+			Help:      "The durations saving data to storage in DataNode.",
+			Buckets:   []float64{0, 10, 100, 200, 400, 1000, 10000},
+		}, []string{
+			"save_duration",
+			nodeIDLabelName,
+		})
+
+	DataNodeFlushCount = prometheus.NewCounterVec(
```
Comment (czs007) [inline internal/metrics/datanode.go:188]: Rename to DataNodeFlushReqCount

File: internal/metrics/datanode.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+var (
+	DataNodeNumFlowGraphs = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_flow_graphs",
+			Help:      "Number of flow graphs in DataNode.",
+		}, []string{
+			"num_flow_graphs",
+			nodeIDLabelName,
```
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/metrics/datanode.go:62]: fixed

File: internal/metrics/datanode.go (hunk 1 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:70]: fixed

File: internal/metrics/datanode.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+		})
+
+	DataNodeNumDmlChannels = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: milvusNamespace,
+			Subsystem: typeutil.DataNodeRole,
+			Name:      "num_dml_channels",
+			Help:      "Number of dmlChannels per collection in DataNode.",
+		}, []string{
+			"num_dml_channels",
```
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/metrics/datanode.go:94]: fixed

File: internal/metrics/datanode.go (hunk 2 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:105]: fixed

File: internal/metrics/datanode.go (hunk 3 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:116]: fixed

File: internal/metrics/datanode.go (hunk 4 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:125]: fixed

File: internal/metrics/datanode.go (hunk 5 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:138]: fixed

File: internal/metrics/datanode.go (hunk 6 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:164]: arguably

File: internal/metrics/datanode.go (hunk 7 above)
Comment (bigsheeper) [reply internal/metrics/datanode.go:188]: arguably

**Review by LoveEachDay — APPROVED** [review]
/lgtm

**Pull request merged**